p2p/src/network/pubsub/p2p_network_pubsub_reducer.rs

```diff
@@ -211,96 +289,292 @@ impl P2pNetworkPubsubState {
                             }
                         }
                         Err(err) => {
                             dbg!(err);
                         }
                     }
+                    }
                 }
+
+                let (dispatcher, global_state) = state_context.into_dispatcher_and_state();
+                let state: &Self = global_state.substate()?;
+                let config: &P2pConfig = global_state.substate()?;
+                let peer_id = *peer_id;
+
+                let incoming_block = state.incoming_block.as_ref().cloned();
+                let incoming_transactions = state.incoming_transactions.clone();
+                let incoming_snarks = state.incoming_snarks.clone();
+                let topics = state.topics.clone();
+
+                let mut prune_at_topics = vec![];
+                for (topic_id, map) in topics {
+                    let mesh_size = map.values().filter(|s| s.on_mesh()).count();
+                    let could_accept = mesh_size < config.meshsub.outbound_degree_high;
+
+                    if !could_accept {
+                        if let Some(topic_state) = map.get(&peer_id) {
+                            if topic_state.on_mesh() {
+                                prune_at_topics.push(topic_id);
+                            }
+                        }
+                    }
+                }
+
+                for topic_id in prune_at_topics {
+                    dispatcher.push(P2pNetworkPubsubAction::Prune { peer_id, topic_id });
+                }
+
+                broadcast(dispatcher, global_state)?;
+                if let Some((_, block)) = incoming_block {
+                    let best_tip = BlockWithHash::new(Arc::new(block));
+                    dispatcher.push(P2pPeerAction::BestTipUpdate { peer_id, best_tip });
+                }
+                for (transaction, nonce) in incoming_transactions {
+                    dispatcher.push(P2pChannelsTransactionAction::Libp2pReceived {
+                        peer_id,
+                        transaction: Box::new(transaction),
+                        nonce,
+                    });
+                }
+                for (snark, nonce) in incoming_snarks {
+                    dispatcher.push(P2pChannelsSnarkAction::Libp2pReceived {
+                        peer_id,
+                        snark: Box::new(snark),
+                        nonce,
+                    });
+                }
+                Ok(())
             }
             // we want to add peer to our mesh
             P2pNetworkPubsubAction::Graft { peer_id, topic_id } => {
-                let Some(state) = self
+                let Some(state) = pubsub_state
                     .topics
                     .get_mut(topic_id)
                     .and_then(|m| m.get_mut(peer_id))
                 else {
-                    return;
+                    return Ok(());
```

Comment:
is this a bug condition?

Reply:
This is not bug condition it can happen.